[WIP] TextualSummaries: part of client side rendering with React.

martinpovolny opened the pull request:
The goal is to implement a top level component for client-side rendering of textual summaries.

### TODO
 * tests (snapshot testing)
 * more formatting variant
 * links

why is the decostruction not working? `(item, i) => FooBar({key: i, ...item})`

ohadlevy commented on app/javascript/react/textual_summary/icon_or_image.jsx:
```diff
@@ -0,0 +1,42 @@
+  image: null,
+  icon: null,
+};
+
```
can be written as:
```js
import * as React from 'react';
import PropTypes from 'prop-types';

/**
 * Render icon or an image with a title
 */
const IconOrImage = ({ icon, title, image }) => {
  const renderIcon = () => <i className={icon} />;

  const renderImage = () => <img src={image} alt={title} title={title} />;

  return image ? renderIcon() : renderImage();
};

IconOrImage.propTypes = {
  title: PropTypes.string,
  image: PropTypes.string,
  icon: PropTypes.string,
};

IconOrImage.defaultProps = {
  title: null,
  image: null,
  icon: null,
};

export default IconOrImage;
```

martinpovolny replied:
yes, I will move the methods inside the class (I did that in some places already).

But the line `return image ? renderIcon() : renderImage();` is not the same as my original code. This simplification would not work.

ohadlevy replied:
sorry i did not test the code, what did not work?  how about
```js
return (image === null ?  renderIcon() : renderImage());
```

martinpovolny replied:
no ;-)

in case of both not present I need an empty string. And nested ternary is discouraged so I have that `if`

ohadlevy commented on package.json:
```diff
@@ -54,6 +54,8 @@
     "babel-preset-env": "~1.4.0",
     "babel-preset-react": "^6.24.1",
+    "babel-preset-es2015": "^6.24.1",
+    "babel-preset-stage-1": "^6.24.1",
```
are those required?

martinpovolny replied:
```
const renderIcon = () => <i className={icon} />;
```
Yes. W/o those 2 lines this code would not compile ^^^

ohadlevy replied:
what was the error? stage-1 should normally not be required.

martinpovolny replied:
```
    at transpile (/home/martin/Projects/manageiq-ui-classic/node_modules/babel-loader/lib/index.js:63:13)
    at Object.module.exports (/home/martin/Projects/manageiq-ui-classic/node_modules/babel-loader/lib/index.js:163:20)
 @ ./app/javascript/react/textual_summary/textual_group.jsx 20:17-39
 @ ./app/javascript/react/textual_summary/textual_row.jsx
 @ ./app/javascript/react/textual_summary/index.jsx
 @ ./app/javascript/packs/application-common.js

ERROR in ./app/javascript/react/textual_summary/icon_or_image.jsx
Module build failed: SyntaxError: Unexpected token (12:13)

  10 |   }
  11 | 
> 12 |   renderIcon = () => <i className={this.props.icon} title={this.props.title} />;
     |              ^

```

```
diff --git a/.babelrc b/.babelrc
index 9d3ea8b..936c5c8 100644
--- a/.babelrc
+++ b/.babelrc
@@ -1,5 +1,6 @@
 {
-  "presets": [["env", { "modules": false }], "react", "es2015", "stage-1"],
+  //"presets": [["env", { "modules": false }], "react", "es2015", "stage-1"],
+  "presets": [["env", { "modules": false }], "react", "es2015"],
   "env": {
     "test": {
       "plugins": ["transform-es2015-modules-commonjs"]
```

martinpovolny replied:
The problem is it's inside the class, so I need class fields. I can change the code to use methods instead or make it a function, not a class.

himdel replied:
The missing feature is called class fields: https://github.com/tc39/proposal-class-fields

```
class Foo {
  renderIcon = 5;
}
```

Currently at stage3 ..  so maybe we just need to bump es2015 to a newer version..

himdel replied:
Looks like maybe we should just be using `env` (and `react`) with some settings to target IE9+.

martinpovolny replied:
We need to figure the best way ;-)

Figuring the best way, I have some notes:

I like some of the poweful expressivity of the functional notation. But there are some parts that are really ugly inside JSX:
 * the contional inclusion of markup: `(condition && (<your_crap_here>))`
```
            {(index === 0) && (                                                                     
              <td rowSpan={item.value.length} className="label">{item.label}</td>                   
            )}   
```

 * case construct (hash + lookup): 
```
      {{                                                                                            
          'shared/summary/textual': (                                                                 
            <GenericGroup items={props.group.items} title={props.group.title} />                      
          ),                                                                                          
          'shared/summary/textual_tags': (                                                            
            <TagGroup items={props.group.items} title={props.group.title} />                          
          ),                                                                                          
          'shared/summary/textual_table': (                                                           
>>          <SimpleTable items={props.group.items} title={props.group.title} />                       
          ),                                                                                          
        }[props.group.partial]} 
```

 * I am glad for the `<React.Fragment>`. That helps.

Overall the JSX seems to be just another way to write markup and not particularly better than HAML. How many do we need?

martinpovolny commented on app/javascript/react/textual_summary/icon_or_image.jsx:
```diff
@@ -0,0 +1,42 @@
+
+  renderIcon() {
+    // %i{:class => item[:icon], :title => item[:title]}
+    return <i className={this.props.icon} title={this.props.title} />;
```
Not now, later when we include patternfly-react in manageiq, we can do it.

ohadlevy replied:
TBH I think we should just add it, no point in writing the code twice...

martinpovolny replied:
"writing the code twice..."  -- you mean the single line ;-).

I will take a look at a later point. It's not a 5 minutes job. After 5 minutes of searching I found the component in patternfly-react. Figured it's not in the storybook https://rawgit.com/patternfly/patternfly-react/gh-pages/index.html. Have to figure what that means, hopefully it does not mean it should not be used.

Then I figured it's actually 2 components: https://github.com/patternfly/patternfly-react/blob/master/src/components/Icon/Icon.js and the resulting markup seems different. So have to check if it renders that same.

All that work is to prevent writing `<i className={this.props.icon} title={this.props.title} />;` Probably not. That does not match my sense of effective work.

In this PR I am trying to start this task by moving the rendering to the client side while making sure that the textual summaries render as they did before. W/o extra changes if possible.

Later we can change the markup and/or styling. In a separate PR or at least a separate commit.

ohadlevy replied:
I've added https://github.com/patternfly/patternfly-react/issues/234 to track the missing storybook.

the two components are simply because we support two type of components (patternfly icons and font-awesome)
using <i> i assume its only font-awesome?

to use it you can simply do
```js
import { Icon } from 'patternfly-react';

...
export default = () => (
  <Icon type='fa' name='react' />
);

martinpovolny commented on app/javascript/react/textual_summary/generic_group.jsx:
```diff
@@ -0,0 +1,30 @@
+import PropTypes from 'prop-types';
+import GenericTableRow from './generic_table_row';
+
+export default class GenericGroup extends React.Component {
```
> this can probably be written as a const,

yes, the ones where there'll be just the `render` and no other code I'll convert to functions. (I guess you mean "use function rather thatn class").

> and probably consider using patternfly-react table component.

not in the repo at this point, but yes, I will take a look if it's usable in this context

martinpovolny commented on app/javascript/packs/application-common.js:
```diff
@@ -5,11 +5,21 @@
+
+$(function () {
+  const summary = $('#textual_summary').data('summary');
+  ReactDOM.render(<TextualSummary summary={summary} />, document.getElementById('textual_summary'));
```
yes, will do that

martinpovolny replied:
cannot be done right now, unless the component uses only `data`:

https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/javascript/react/componentRegistry.js#L43

will wait with this change until that is fixed

himdel replied:
Should be resolved in https://github.com/ManageIQ/manageiq-ui-classic/pull/3228, reviewing tomorrow..

martinpovolny commented on app/javascript/react/textual_summary/generic_group.jsx:
```diff
@@ -0,0 +1,30 @@
+import * as React from 'react';
```
Well. I am not sure if this is a good idea. @himdel, @karelhala, @skateman, @Hyperkid123: what do you think?

karelhala replied:
To be honest I like to use `.jsx` for react specific code, that way you can immediately see that it's react component and nothing else.

I would even recommend using this `${cmp_name}.component.jsx` for components, `${action_name}.action.js` for redux actions and `${reducer_name}.reducer.js` for reducers, but that is my preference and no need to do it that way.

himdel replied:
Same here, better explicit than implicit :)

ohadlevy replied:
I disagree :) but thats OK..  there is an ongoing discussion in the foreman
community about how to structure this (and redux etc) at
theforeman/foreman#5240

Hyperkid123 replied:
If we were using only React i would be fine with `.js`, but i think it would be better to use `.jsx`.

ohadlevy commented:
On Mon, Feb 19, 2018 at 5:20 PM, Karel Hala <notifications@github.com>
wrote:

> *@karelhala* commented on this pull request.
> ------------------------------
>
> In app/javascript/react/textual_summary/generic_group.jsx
> <https://github.com/ManageIQ/manageiq-ui-classic/pull/3420#discussion_r169105113>
> :
>
> > @@ -0,0 +1,30 @@
> +import * as React from 'react';
>
> To be honest I like to use .jsx for react specific code, that way you can
> immediately see that it's react component and nothing else.
>
> I would even recommend using this ${cmp_name}.component.jsx for
> components, ${action_name}.action.js for redux actions and
> ${reducer_name}.reducer.js for reducers, but that is my preference and no
> need to do it that way.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ManageIQ/manageiq-ui-classic/pull/3420#discussion_r169105113>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx9SE7Y9TdYvMAVsNC0QU3AwKo8_qks5tWZFBgaJpZM4SI4mV>
> .
>

I disagree :) but thats OK..  there is an ongoing discussion in the foreman
community about how to structure this (and redux etc) at
https://github.com/theforeman/foreman/pull/5240

martinpovolny commented:
@AparnaKarve : can you, please, tell me, why the `:no_value` key was added in 
 https://github.com/ManageIQ/manageiq-ui-classic/pull/2852 ? Thx!

AparnaKarve commented:
`no_value` is to render a table row with a single cell with no associated column or value.
So far it has been used only in the Generic Objects context.

In the image below, notice the Attributes, Associations and Methods table, that do not have a value.
All we need is a single cell that displays - No Attributes defined

<img width="1422" alt="screen shot 2018-02-28 at 4 16 44 pm" src="https://user-images.githubusercontent.com/1538216/36820459-d1f4023e-1ca2-11e8-872b-9d6b4181db3f.png">

martinpovolny commented:
> no_value is to render a table row with a single cell with no associated column or value.
So far it has been used only in the Generic Objects context.

Thanks, @AparnaKarve. I saw that much it the code ;-) But I don't understand why we need a separate attribute `:no_value`. Why don't we just check `:value` for emptyness. Can you explain that, please?

Thanks!

martinpovolny commented:
@AparnaKarve : to be more explicit: What was the motivation behind the PR that introduced the new attribute? Why testing just `:value` was not enough?

I'd like to remove this feature if it's not really needed, but there's no test that would fail if I do nor do I understand it when reading the PR.

Thx!

himdel commented:
@martinpovolny the babel stuff .. 

```diff
diff --git a/.babelrc b/.babelrc
index 9d3ea8b30..83a481bc7 100644
--- a/.babelrc
+++ b/.babelrc
@@ -1,5 +1,6 @@
 {
-  "presets": [["env", { "modules": false }], "react", "es2015", "stage-1"],
+  "presets": [["env", { "modules": false }], "react"],
+  "plugins": ["transform-class-properties", "transform-object-rest-spread"],
   "env": {
     "test": {
       "plugins": ["transform-es2015-modules-commonjs"]
diff --git a/package.json b/package.json
index 394df1c49..77e5ed098 100644
--- a/package.json
+++ b/package.json
@@ -54,12 +54,11 @@
     "babel-jest": "^22.0.0",
     "babel-loader": "~7.0",
     "babel-plugin-syntax-dynamic-import": "^6.18.0",
-    "babel-plugin-transform-class-properties": "^6.24.1",
+    "babel-plugin-transform-class-properties": "~6.24.1",
+    "babel-plugin-transform-object-rest-spread": "~6.26.0",
     "babel-polyfill": "^6.23.0",
-    "babel-preset-env": "~1.4.0",
-    "babel-preset-es2015": "^6.24.1",
+    "babel-preset-env": "^1.6.1",
     "babel-preset-react": "^6.24.1",
-    "babel-preset-stage-1": "^6.24.1",
     "coffee-loader": "~0.7.3",
     "coffee-script": "~1.12.5",
     "compression-webpack-plugin": "~0.4.0",
```

AparnaKarve commented:
> Why don't we just check :value for emptyness

That's what I thought too and even adjusted the code to check for existence of `:value` 
https://github.com/ManageIQ/manageiq-ui-classic/pull/2846/commits/5e2b5a660c3b866d921eca29eab3173cec93dd63 (PR #2846)

But turns out that a nil value or a blank value could be a valid value in certain cases - in which case we need to render the value cell.
(I think the Foreman configured systems summary screen has such examples)

martinpovolny commented:
@AparnaKarve : Can we have a nil value render nothing and an empty string to be rendered as empty string?

The 5e2b5a6 that you referenced is what turned out not to work? If so then it is pretty counter-intuitive that it does not work I would suggest having it documented in a test or at least add a source code comment. Can you remember what the problem was?

Taking it from the other side: why cannot the GO behave the same way, why is GO the only entity that needs a special behavior? Would it be a problem if the behavior was consistent?

I mean we have 2 behaviors for empty values:
  1. empty groups are not rendered at all so instead of a header "Associations" and text "No Associations defined" there would simply be nothing. That would be consistent with all the other summary pages.

  2. The helper functions that render links to association also can handle a link to an empty list.


For some reason, it seems that you implemented a 3rd way:
```
  def textual_group_associations                                                                    
    if @record.property_associations.count > 0                                                      
      TextualGroup.new(_("Associations"), associations)                                             
    else                                                                                            
      TextualGroup.new(_("Associations"), %i(associations_none))                                    
    end                                                                                             
  end 
```

So this is completely inconsistend with all the other textual summaries. It's a new type of behavior.

If I accept that then looking at the code path that one has to go to display:
```
+----------------------------------------+
| Associations...........................|
+----------------------------------------+
| No Associations defined................|
+----------------------------------------+
```
Is pretty obscure. I mean the functionality to display this is added to the most complex path in the TextualSummary rendering.


#### So what next?

I suggest that:
 1. I remove the functionality.
 1. You then help me understand what where the requirements or reasons behind the different behavior of GO compared to all the other places.
 1. If it really cannot behave the same as the other places, I get it working again.
 1. or you show me the places that get broken (somewhere under Foreman)

This way we should end up with less code, more tests and more consistend UX.

martinpovolny commented:
@AparnaKarve : I also need you help with this: https://github.com/ManageIQ/manageiq-ui-classic/issues/3492

It's another area where the behavior is different to the other summaries. It's hard for me to fix that, because I have no data for the physical infra. Given you reviewed it, can you show me how you tested/debugged? I mean do you have a gist or script or database, that I could use? I don't see those two special partials covered in any tests so I need a some help.

AparnaKarve commented:
> I remove the functionality.
You then help me understand what where the requirements or reasons behind the different behavior of GO compared to all the other places.

GO instances need to display Attributes, Associations and Methods.
If these don't exist, then we need to **explicitly** say so.

Not showing data because it does not exist can be confusing for the user - I think that's what the BZ was about.
The `No Attributes defined` text clearly conveys that there are no Attributes defined for the GO instance.

> or you show me the places that get broken (somewhere under Foreman)

With https://github.com/ManageIQ/manageiq-ui-classic/commit/5e2b5a660c3b866d921eca29eab3173cec93dd63, certain tables in Foreman Configured Systems summary page looked like this without the value column, which is incorrect.

(Bad)
<img width="494" alt="screen shot 2018-03-02 at 4 16 39 pm" src="https://user-images.githubusercontent.com/1538216/36927891-a5c1de40-1e35-11e8-883e-272c6894a9cf.png">

Compare the above with this -

(Good)
<img width="497" alt="screen shot 2018-03-02 at 4 23 37 pm" src="https://user-images.githubusercontent.com/1538216/36927958-24603f58-1e36-11e8-8b8f-efdb1c4f43de.png">

----------
> because I have no data for the physical infra

6_0_dc_MBU_demo_01202018.backup from Dan's folder should have all the data for physical infra.
Look up record# 55 (/physical_server/show/55)

martinpovolny commented:
@AparnaKarve : I have redone the empty groups a bit. I created a trivial component to display those, rather than having that behavior mixed onto the complex table.

I found that the behavior is different in GO vs GO definitions. A am not sure that is desired. In my eyes it's yet another inconsistency.

I did some extra fixing:
  * I fixed the broken i18n, (but meanwhile @mzazrivec fixed that in master so I effectively removed those changes when rebasing).
 * I removed some methods, 2 where broken anyway (returning identity where `each` was used in place of `map`).

Overall it should work better that before and your review is welcome.

AparnaKarve commented:
> I created a trivial component to display those

Thanks. I was going to suggest that as well.

> the behavior is different in GO vs GO definitions

Non-admin users cannot view a GO definition.
Hence for them, information like "No Attributes/Associations/Methods defined" can be quite useful.

Admins, on the other hand, can view the GO definitions.

However, for the sake of consistency, we could also add "No Attributes/Associations/Methods defined" for GO definitions.

martinpovolny commented:
@AparnaKarve : I figured one of the Physical summary files is unused and the 2nd can be trivially replaced with a generic one: Please, check the 2 that mentin "physical" close to the end.
p.s. It should also fix the i18n in the summary.

AparnaKarve commented:
The `TextualTable` implementation looks nice and clean. 

The unused one is a result of this PR https://github.com/ManageIQ/manageiq-ui-classic/pull/3398.
The network adapters (aka guest devices) now have a view of their own, which means they don't need to be displayed in the Physical Server Summary screen anymore.

martinpovolny commented:
@dclarizio : a note regarding client-side PDF printing.

I tested https://github.com/MrRio/jsPDF and immediately run into https://github.com/MrRio/jsPDF/issues/1434 (colspan and rowspan don't work).

skateman commented:
When you open an openstack computenode's (infra -> hosts) summary screen, the boolean value isn't properly displayed for the introspected field:
**Before:**
![screenshot from 2018-05-24 14-35-23](https://user-images.githubusercontent.com/649130/40485764-b884b4f6-5f5f-11e8-8679-562632de552e.png)
**After:**
![screenshot from 2018-05-24 14-34-58](https://user-images.githubusercontent.com/649130/40485737-a892d5d2-5f5f-11e8-938f-508df2c716d5.png)

skateman commented:
With configuration profiles this one:
![screenshot from 2018-05-24 15-38-51](https://user-images.githubusercontent.com/649130/40489074-96543baa-5f68-11e8-98f1-625656b1ef02.png)

skateman commented:
And configured systems:
![screenshot from 2018-05-24 15-40-12](https://user-images.githubusercontent.com/649130/40489162-cd4aeabe-5f68-11e8-8460-b79787246ed7.png)

martinpovolny commented:
ping @mzazrivec, @himdel : please, merge

(this is going to break already broken printing even more, but that must be fixed separately)

skateman commented:
I'm working on the printing right now, let's get this in so I can test it properly :wink:
